### ISSUE TITLE
Support standalone mode (non-cluster)

### ISSUE THREAD

**Issue by zuiderkwast** [Ericsson/ered#27]
We should change the API to make it friendly for standalone instances and not only clusters:

* `ered:connect(Host, Port, Opts)` for standalone mode, a single connection to a single node.
* `ered:connect_cluster(AddressesAndPorts, Opts)` for cluster.
* An ered handle can be represented as a small tuple such as `{cluster, Pid}`. This is similar to how a `gen_tcp` or `ssl` handle is represented.
* Functions like `ered:command(EredRef, Command)` can work for cluster and standalone and determine by the ered handle (the small tuple) if it's a cluster instance or a single connection.

#41 should be merged first.

Should we autodetect cluster/standalone mode? We could, for example handle an error response form CLUSTER SLOTS. This is not currently the plan though. Users can explicitly connect to a cluster or a standalone node.

**Comment by ioolkos** [comment]
@zuiderkwast is it possible to connect to a standalone node, with PR https://github.com/Ericsson/ered/pull/41 ?

**Comment by zuiderkwast** [comment]
Even without #41, it's possible to connect to a single node using the `ered_client` module. #41 does make things a little more clear though with functions like ered:connect_client for connecting to a single instance.

It's not possible to use a standalone instance as if it were a cluster though, with the same functions, but actually I don't think we want to do this automatically. Cluster and standalone deployments are very different. Standalone mode allows SELECT and cross-slot commands, for example, which are forbidden in cluster mode.

**Comment by ioolkos** [comment]
@zuiderkwast Thanks a lot for your useful information! Currently we're using eredis in VerneMQ so we cannot connect to a Redis cluster yet. And ered's stated goal is to replace both previous libraries, isn't it. In any case, I'll try to do some testing based on your suggestions.

**Comment by zuiderkwast** [comment]
Yes, ered implements the functionality for single instance and cluster, but the API has a preference for cluster. This was done before I was the maintainer of this project. Now, I've been trying to make the API look nice for single instance too.

An idea: We use a tuple like `{cluster, Pid}` or `{standalone, Pid}` as the ered instance reference, returned by `ered:connect` functions. Then we can use the same functions like `ered:command/N` for both of them and it can dispatch to the right module based on the tuple. That's future work, if you like the idea. (`ssl:connect` and `gen_tcp:connect` return a tuples like this, so it's not that strange.)

I've been busy with other work though.

**Comment by ioolkos** [comment]
@zuiderkwast I suppose the client always knows whether it's connecting to a standalone or a cluster. Your suggestion sounds great to me, in any case!

**Comment by ioolkos** [comment]
I've now written my plugin using `ered:start_link/2` as main entry point. As expected, this only works for a cluster, not a standalone (standalone gets the `unmapped_slots` error). A question, though: with the idea proposed above, will `ered` be able to figure out by itself whether it connects to a standalone or a cluster?

**Comment by zuiderkwast** [comment]
> A question, though: with the idea proposed above, will `ered` be able to figure out by itself whether it connects to a standalone or a cluster?

It would be possible to auto-detect, but I think it's better that users connect explicitly to cluster or standalone. I'm thinking `ered:connect(Host, Port, Opts)` for standalone and `ered:connect_cluster(AddressesAndPorts, Opts)` for cluster. After that, `ered:command(EredRef, Command)` and similar can be used for both kinds of instances.

**Comment by ioolkos** [comment]
Sounds great. That interface is perfect at least for me, as the client (the plugin) always knows whether it connects to a cluster or not.
Let me know if I can test  (or do anything helpful). No worry & no hurry for anyone :)
I think it'll be a cool library to use and like the unified approach in it.
